src/diagnostics.rs

```diff
@@ -159,47 +192,110 @@ impl Diagnostics {
   /// Returns a Display that prints the diagnostics with one of the given severities.
   pub fn display_with_severity(&self, severity: Severity) -> impl std::fmt::Display + '_ {
     DisplayFn(move |f| {
-      fn filter<'a>(
-        errs: impl IntoIterator<Item = &'a Diagnostic>,
-        severity: Severity,
-      ) -> impl Iterator<Item = &'a Diagnostic> {
-        errs.into_iter().filter(move |err| err.severity == severity)
-      }
+      // We want to print diagnostics information somewhat like this:
+      // ```
+      // In file A :
+      // In definition X :
+      //   {error}
+      // In definition Y :
+      //   {error}
+      //
+      // In file B :
+      // In compiled Inet Z :
+      //   {error}
+      //
+      // Other diagnostics:
+      // In {...}
+      // ```
+      // The problem is, diagnostics data is currently structured as a mapping from something like
+      // DiagnosticOrigin to Vec<(DiagnosticMessage, DiagnosticFile)>, and we would need something
+      // like a mapping from DiagnosticFile to DiagnosticOrigin to Vec<DiagnosticMessage> in order
+      // to print it cleanly. We might want to change it later to have this structure,
+      // but meanwhile, we do the transformations below to make the goal possible.
+
+      // Ignore diagnostics without the desired severity.
+      let diagnostics = self
+        .diagnostics
+        .iter()
+        .map(|(origin, diags)| (origin, diags.iter().filter(|diag| diag.severity == severity)));
 
-      let mut has_msg = false;
-      for (orig, errs) in &self.diagnostics {
-        let mut errs = filter(errs, severity).peekable();
-        if errs.peek().is_some() {
-          match orig {
-            DiagnosticOrigin::Book => {
-              for err in errs {
-                writeln!(f, "{err}")?;
+      // Produce the structure described above.
+      let groups: BTreeMap<&Option<String>, BTreeMap<&DiagnosticOrigin, Vec<&Diagnostic>>> = diagnostics
+        .fold(BTreeMap::new(), |mut file_tree, (origin, diags)| {
+          for diag in diags {
+            // We need to allow this Clippy warning due to `Name` in `DiagnosticOrigin::Function`.
+            // We know how it works, so it shouldn't be a problem.
+            #[allow(clippy::mutable_key_type)]
```

Comment:
We can allow it at the crate level, since we're just sprinkling this macro in different places, always for the same reason.